lib/cstruct.ml

```diff
@@ -254,41 +255,869 @@ external swap64 : int64 -> int64 = "%bswap_int64"
+         * and bigger than a granularity unit (byte) *)
+        else
+            (* Check if it crosses the boundary *)
+            if i mod align = 0 then
+                (* TODO: Consider also the length of the buffer here *)
+                let count = i / align in
+                if count <= 2 then get_uint16_chop_boundary swap p t byteoff
+                else
+                    (* For now just throw an exception *)
+                    err_invalid_bounds "get_bits_uint8" t i 2
+            else
+                let count = i / align + 1 in
+                (* Here mask is tricky - what to read there?
+                 * Do we need to make swap before or after? *)
+                let mask = (1 lsl i) - 1 in
+                let raw =
+                    if count <= 2 then get_uint16_chop_boundary swap p t byteoff
+                    else
+                        (* For now just throuw an exception *)
+                        err_invalid_bounds "get_bits_uint8" t i 2
+                in
+                raw land mask
+      (* When they start not at granularity boundary *)
+      else
+          let bithead = o mod align in
+          (* Third case - when they are located not at a granularity boundary,
+           * and lesser or equal than a tail of the granula (byte) *)
+          if i + bithead < align then
+            let byte = get_uint8 t byteoff in
+            let mask = ((1 lsl i) - 1) lsl bithead in
+            (byte land mask) lsr bithead
+          (* Fouth case - when they are located not at a granularity boundary,
+           * and go further than next granularity boundary *)
+          else
+            (* Check if it crosses the boundary *)
+            if (i + o) mod align = 0 then
+                let count = i / align in
+                (* Read these bytes, how to check the swap? *)
+                let mask = ((1 lsl i) - 1) lsl bithead in
+                let raw =
+                    if count <= 2 then get_uint16_chop_boundary swap p t byteoff
+                    else
+                        (* For now just throw an exception *)
+                        err_invalid_bounds "get_bits_uint8" t i 2
+                in
+                (raw land mask) lsr bithead
+            else
+                let count = i / align + 1 in
+                (* Read these bytes, how to check the swap? *)
+                let mask = ((1 lsl i) - 1) lsl bithead in
+                let raw =
+                    if count <= 2 then get_uint16_chop_boundary swap p t byteoff
+                    else
+                        (* For now just throw an exception *)
+                        err_invalid_bounds "get_bits_uint8" t i 2
+                in
+                (raw land mask) lsr bithead
+
+(* FIXME: Do not tie up itself to the byte as granularity unit *)
+(* This function cannot read more than 16 bits *)
+let get_bits_uint16_le swap p t (o:int) (i:int) =
+    Printf.printf "get_bits_uint16 [%x] bits @ %x\n" i o;
```

Comment:
Should be removed?

Reply:
Yes, of course I remove this, once it is more or less ready. Along with comments cleanup. I want more general concept review first, so I know where to change/fix/etc.